src/realizations/catchment/Bmi_Multi_Formulation.cpp

```diff
@@ -0,0 +1,266 @@
+
+        for (const std::string &var_name : module->get_bmi_input_variables()) {
+            std::string framework_alias = module->get_config_mapped_variable_name(var_name);
+            (*var_aliases)[framework_alias] = var_name;
+            // If framework_name is not in collection from which we have available data sources ...
+            if (availableData.count(framework_alias) != 1) {
+                throw std::runtime_error(
+                        "Multi BMI cannot be created with module " + module->get_model_type_name() + " with input " +
+                        "variable " + framework_alias +
+                        (var_name == framework_alias ? "" : " (an alias of BMI variable " + var_name + ")") +
+                        " when there is no previously-enabled source for this input.");
+            }
+            else {
+                module->input_forcing_providers[var_name] = availableData[framework_alias];
+                module->input_forcing_providers[framework_alias] = availableData[framework_alias];
+            }
+        }
+        // Also add the output variable aliases
+        for (const std::string &var_name : module->get_bmi_output_variables()) {
+            std::string framework_alias = module->get_config_mapped_variable_name(var_name);
+            (*var_aliases)[framework_alias] = var_name;
+            if (availableData.count(framework_alias) > 0) {
+                throw std::runtime_error(
+                        "Multi BMI cannot be created with module " + module->get_model_type_name() +
+                        " with output variable " + framework_alias +
+                        (var_name == framework_alias ? "" : " (an alias of BMI variable " + var_name + ")") +
+                        " because a previous module is using this output variable name/alias.");
+            }
+            availableData[framework_alias] = module;
+        }
+        module_variable_maps[i] = var_aliases;
+    }
+
+    // TODO: get synced start_time values for all models
+    // TODO: get synced end_time values for all models
+
+    // TODO: set up output_variables (add support for controlling the particular sub-formulation somehow)
+
+    // TODO: account for consistent setting of 'allow_exceed_end_time'
+
+    // Output header fields, if present
+    auto out_headers_it = properties.find(BMI_REALIZATION_CFG_PARAM_OPT__OUT_HEADER_FIELDS);
+    if (out_headers_it != properties.end()) {
+        std::vector<geojson::JSONProperty> out_headers_json_list = out_headers_it->second.as_list();
+        std::vector<std::string> out_headers(out_headers_json_list.size());
+        for (int i = 0; i < out_headers_json_list.size(); ++i) {
+            out_headers[i] = out_headers_json_list[i].as_string();
+        }
+        set_output_header_fields(out_headers);
+    }
+    else {
+        set_output_header_fields(get_output_variable_names());
+    }
+}
+
+const bool &Bmi_Multi_Formulation::get_allow_model_exceed_end_time() const {
```

Comment:
I would add a comment (even if redundant with the header) here describing the semantics of `allow_model_exceed_end_time` when it may not be the same across each module.